tests/test_lists2safebrowsing.py

```diff
@@ -97,3 +99,62 @@ def test_canonicalize(url, expected):
     documentation with some adjustments and additions.
     """
     assert canonicalize(url) == expected
+
+
+def _aux_add_domain_to_list(domain, previous_domains=set()):
+    """Auxiliary function for add_domain_to_list tests."""
+    canonicalized_domain = canonicalize(domain)
+    domain_hash = hashlib.sha256(canonicalized_domain)
+    output = []
+
+    with patch("test_lists2safebrowsing.open", mock_open()):
+        with open("test_blocklist.log", "w") as log_file:
+            added = add_domain_to_list(domain, previous_domains, set(),
+                                       log_file, output)
+            log_writes = log_file.write.call_args_list
+
+    return (added, canonicalized_domain, domain_hash, previous_domains,
+            log_writes, output)
+
+
+def test_add_domain_to_list():
+    """Test adding a domain to a blocklist."""
+    domain = "https://www.host.com"
+    (added, canonicalized_domain, domain_hash, previous_domains,
+     log_writes, output) = _aux_add_domain_to_list(domain)
+
+    expected_log_writes = [
+        call("[m] %s >> %s\n" % (domain, canonicalized_domain)),
+        call("[canonicalized] %s\n" % canonicalized_domain),
+        call("[hash] %s\n" % domain_hash.hexdigest()),
+    ]
+
+    assert added
+    assert canonicalized_domain in previous_domains
+    assert domain_hash.digest() in output
+    assert log_writes == expected_log_writes
+
+
+def test_add_domain_to_list_public_suffix():
+    """Test that add_domain_to_list skips public suffix domains."""
+    domain = "https://co.uk"
+    (added, canonicalized_domain, domain_hash, previous_domains,
+     log_writes, output) = _aux_add_domain_to_list(domain)
+
+    expected_log_writes = [call("[Public Suffix] %s; Skipping.\n" %
+                           canonicalized_domain.rstrip("/"))]
+
+    assert not added
+    assert canonicalized_domain not in previous_domains
+    assert domain_hash.digest() not in output
+    assert log_writes == expected_log_writes
+
+
+def test_add_domain_to_list_duplicate(tmpdir):
+    """Test that add_domain_to_list does not add domains twice."""
+    domain = "https://duplicate.com"
+    (added, _, domain_hash, _, log_writes, output) = \
+        _aux_add_domain_to_list(domain, {canonicalize(domain)})
+    assert not added
+    assert domain_hash.digest() not in output
+    assert not log_writes
```

Comment:
This is a failing test case. I would like it to either: (1) the `output` matches what the `previous_domains` is (if the `previous_domain` as abc.com the `output` should also have hashed abc.com) OR we call the `add_domain_to_list` twice on this test and assert twice (first time the domain is added but the second time the domain should not be added and the test fails at this point because we have a bug in the`add_domain_to_list`)

It's ok to commit to master with a failing test as we will be making a follow-up PR to fix this.

Reply:
Done! I opted to call `add_domain_to_list` twice, as it seemed more intuitive.